scripts/west_commands/runners/blackmagicprobe.py

```diff
@@ -5,11 +5,59 @@
 '''Runner for flashing with Black Magic Probe.'''
 # https://github.com/blacksphere/blackmagic/wiki
 
+import os
 import signal
+import sys
+import glob
 from pathlib import Path
 
 from runners.core import ZephyrBinaryRunner, RunnerCaps
 
+# Default path for linux, based on the project udev.rules file.
+DEFAULT_LINUX_BMP_PATH = '/dev/ttyBmpGdb'
+
+LINUX_SERIAL_GLOB = '/dev/ttyACM*'
+DARWIN_SERIAL_GLOB = '/dev/cu.usbmodem*'
+
+def BlackMagicProbeGDBSerial(port):
+    '''Guess the GDB port for the probe.
+
+    Return the port to use, in order of priority:
+        - the port specified manually
+        - the port in the BMP_GDB_SERIAL environment variable
+        - a guessed one depending on the host
+    '''
+    if port:
+        return port
+
+    if 'BMP_GDB_SERIAL' in os.environ:
+        return os.environ['BMP_GDB_SERIAL']
+
+    platform = sys.platform
+    if platform.startswith('linux'):
+        if os.path.exists(DEFAULT_LINUX_BMP_PATH):
+            return DEFAULT_LINUX_BMP_PATH
+
+        ports = glob.glob(LINUX_SERIAL_GLOB)
+        if not ports:
+            raise RuntimeError(
+                    f'cannot find any valid port matching {LINUX_SERIAL_GLOB}')
+        return sorted(ports)[0]
+
+    elif platform.startswith('darwin'):
+        ports = glob.glob(DARWIN_SERIAL_GLOB)
+        if not ports:
+            raise RuntimeError(
+                    f'cannot find any valid port matching {DARWIN_SERIAL_GLOB}')
+        return sorted(ports)[0]
+
+    elif platform.startswith('win32'):
+        return 'COM1'
```

Comment:
I get that there are some nice udev rules in the linux case, but it seems like this could be made better on Windows and macOS using pyserial.

```
In [1]: import serial.tools.list_ports

In [2]: serial.tools.list_ports.comports(include_links=True)
Out[2]: 
[<serial.tools.list_ports_linux.SysFS at 0x7fe5abb2c700>,
 <serial.tools.list_ports_linux.SysFS at 0x7fe5abb2e320>,
 <serial.tools.list_ports_linux.SysFS at 0x7fe5abb2c430>]

In [3]: [vars(info) for info in serial.tools.list_ports.comports(include_links=True)]
Out[3]: 
[{'device': '/dev/ttyS1',
  'name': 'ttyS1',
  'description': 'ttyS1',
  'hwid': 'PNP0501',
  'vid': None,
  'pid': None,
  'serial_number': None,
  'location': None,
  'manufacturer': None,
  'product': None,
  'interface': None,
  'usb_device_path': None,
  'device_path': '/sys/devices/pnp0/00:05',
  'subsystem': 'pnp',
  'usb_interface_path': None},
 {'device': '/dev/ttyS0',
  'name': 'ttyS0',
  'description': 'ttyS0',
  'hwid': 'PNP0501',
  'vid': None,
  'pid': None,
  'serial_number': None,
  'location': None,
  'manufacturer': None,
  'product': None,
  'interface': None,
  'usb_device_path': None,
  'device_path': '/sys/devices/pnp0/00:04',
  'subsystem': 'pnp',
  'usb_interface_path': None},
 {'device': '/dev/ttyACM0',
  'name': 'ttyACM0',
  'description': 'J-Link - CDC',
  'hwid': 'USB VID:PID=1366:0105 SER=000683441799 LOCATION=3-4.4.4:1.0',
  'vid': 4966,
  'pid': 261,
  'serial_number': '000683441799',
  'location': '3-4.4.4:1.0',
  'manufacturer': 'SEGGER',
  'product': 'J-Link',
  'interface': 'CDC',
  'usb_device_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4',
  'device_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4/3-4.4.4:1.0',
  'subsystem': 'usb',
  'usb_interface_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4/3-4.4.4:1.0'}]
```

Please use the same `MISSING_REQUIREMENTS` pattern as jlink.py for the import, since pyserial is not part of the standard library.

Reply:
I get this on osx:

```
>>> vars(serial.tools.list_ports.comports()[1])
{'device': '/dev/cu.usbmodem93BB879F1', 'name': 'cu.usbmodem93BB879F1', 'description': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'hwid': 'USB VID:PID=1D50:6018 SER=93BB879F LOCATION=0-1', 'vid': 7504, 'pid': 24600, 'serial_number': '93BB879F', 'location': '0-1', 'manufacturer': 'Black Magic Debug', 'product': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'interface': None}
>>> vars(serial.tools.list_ports.comports()[2])
{'device': '/dev/cu.usbmodem93BB879F3', 'name': 'cu.usbmodem93BB879F3', 'description': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'hwid': 'USB VID:PID=1D50:6018 SER=93BB879F LOCATION=0-1', 'vid': 7504, 'pid': 24600, 'serial_number': '93BB879F', 'location': '0-1', 'manufacturer': 'Black Magic Debug', 'product': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'interface': None}
```

No interface unfortunately but it may be ok to filter for product and get the first one. Waiting for an output on windows if you can.

Reply:
actually don't bother, read the code (https://github.com/pyserial/pyserial/blob/31fa4807d73ed4eb9891a88a15817b439c4eea2d/serial/tools/list_ports_windows.py#L332-L334) I think I can rely on vid and pid being there, updated